# Web Scraping Lab

You will find in this notebook some scrapy exercises to practise your scraping skills.

**Tips:**

- Check the response status code for each request to ensure you have obtained the intended contennt.
- Print the response text in each request to understand the kind of info you are getting and its format.
- Check for patterns in the response text to extract the data/info requested in each question.
- Visit each url and take a look at its source through Chrome DevTools. You'll need to identify the html tags, special class names etc. used for the html content you are expected to extract.

- [Requests library](http://docs.python-requests.org/en/master/#the-user-guide) documentation 
- [Beautiful Soup Doc](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
- [Urllib](https://docs.python.org/3/library/urllib.html#module-urllib)
- [re lib](https://docs.python.org/3/library/re.html)
- [lxml lib](https://lxml.de/)
- [Scrapy](https://scrapy.org/)
- [List of HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
- [HTML basics](http://www.simplehtmlguide.com/cheatsheet.php)
- [CSS basics](https://www.cssbasics.com/#page_start)

#### Below are the libraries and modules you may need. `requests`,  `BeautifulSoup` and `pandas` are imported for you. If you prefer to use additional libraries feel free to uncomment them.

In [2]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
# from pprint import pprint
# from lxml import html
# from lxml.html import fromstring
# import urllib.request
# from urllib.request import urlopen
# import random
import re
# import scrapy
from IPython.display import Image, display_html
from IPython.core.display import HTML

#### Download, parse (using BeautifulSoup), and print the content from the Trending Developers page from GitHub:

In [4]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/developers'

In [5]:
#your code
html = requests.get(url).content



#### Display the names of the trending developers retrieved in the previous step.

Your output should be a Python list of developer names. Each name should not contain any html tag.

**Instructions:**

1. Find out the html tag and class names used for the developer names. You can achieve this using Chrome DevTools.

1. Use BeautifulSoup to extract all the html elements that contain the developer names.

1. Use string manipulation techniques to replace whitespaces and linebreaks (i.e. `\n`) in the *text* of each html element. Use a list to store the clean names.

1. Print the list of names.

Your output should look like below:

```
['trimstray (@trimstray)',
 'joewalnes (JoeWalnes)',
 'charlax (Charles-AxelDein)',
 'ForrestKnight (ForrestKnight)',
 'revery-ui (revery-ui)',
 'alibaba (Alibaba)',
 'Microsoft (Microsoft)',
 'github (GitHub)',
 'facebook (Facebook)',
 'boazsegev (Bo)',
 'google (Google)',
 'cloudfetch',
 'sindresorhus (SindreSorhus)',
 'tensorflow',
 'apache (TheApacheSoftwareFoundation)',
 'DevonCrawford (DevonCrawford)',
 'ARMmbed (ArmMbed)',
 'vuejs (vuejs)',
 'fastai (fast.ai)',
 'QiShaoXuan (Qi)',
 'joelparkerhenderson (JoelParkerHenderson)',
 'torvalds (LinusTorvalds)',
 'CyC2018',
 'komeiji-satori (神楽坂覚々)',
 'script-8']
 ```

In [6]:
#your code
html = requests.get(url).content
soup = BeautifulSoup(html, "lxml")

In [7]:
soup

<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8"/>
<link href="https://github.githubassets.com" rel="dns-prefetch"/>
<link href="https://avatars.githubusercontent.com" rel="dns-prefetch"/>
<link href="https://github-cloud.s3.amazonaws.com" rel="dns-prefetch"/>
<link href="https://user-images.githubusercontent.com/" rel="dns-prefetch"/>
<link crossorigin="anonymous" href="https://github.githubassets.com/assets/frameworks-93d34cfdadb1618eb00917161bb7f744.css" integrity="sha512-k9NM/a2xYY6wCRcWG7f3ROm4X5CJNikViGX0N8YIxs6sUYAe/j08/RSHXr3fA9wLIy87AMFCgXm6jbvhZhIXWw==" media="all" rel="stylesheet"/>
<link crossorigin="anonymous" href="https://github.githubassets.com/assets/site-6547fa2bebd0a8c63e461573691982cb.css" integrity="sha512-ZUf6K+vQqMY+RhVzaRmCy2ePbSZad4TkaGRbd6v5gFt6f9Q/nqjDkBDjQgXNmZw7J9mcYxlsE4fhRw7CTluRow==" media="all" rel="stylesheet"/>
<link crossorigin="anonymous" href="https://github.githubassets.com/assets/behaviors-152400ab410cb387f3b645589da1bab5.css" integr

In [12]:
table = soup.find_all('h1',{'class':'h3 lh-condensed'})
print(list(table))
print(len(table))


[<h1 class="h3 lh-condensed">
<a data-hydro-click='{"event_type":"explore.click","payload":{"click_context":"TRENDING_DEVELOPERS_PAGE","click_target":"OWNER","click_visual_representation":"TRENDING_DEVELOPER","actor_id":null,"record_id":439362,"originating_url":"https://github.com/trending/developers","user_id":null}}' data-hydro-click-hmac="f1e20097f0c51fb225c700ce4cdd4b3af39d365b2dd5d1bda119d73295483951" href="/normanmaurer">
            Norman Maurer
</a> </h1>, <h1 class="h3 lh-condensed">
<a data-hydro-click='{"event_type":"explore.click","payload":{"click_context":"TRENDING_DEVELOPERS_PAGE","click_target":"OWNER","click_visual_representation":"TRENDING_DEVELOPER","actor_id":null,"record_id":173661,"originating_url":"https://github.com/trending/developers","user_id":null}}' data-hydro-click-hmac="e5f8b488975347353146deff5d76f57f9c5823c58166af008ff3dade354f7a3a" href="/kripken">
            Alon Zakai
</a> </h1>, <h1 class="h3 lh-condensed">
<a data-hydro-click='{"event_type":"expl

In [24]:
f = lambda x: re.findall("            \w+ \w+|            \w+|            \W.........",str(x) )
u = list(map(f, table ))
sp1 = []
for i in u:
    if i == [] :
        print(i)
    else:
        i = i[-1]
        sp1.append(i)

sp1        
names = []
for i in sp1:
    i = i.strip()
    names.append(i)
    
names    




['Norman Maurer',
 'Alon Zakai',
 'Michele Locati',
 'Ha Thach',
 'Joel Arvidsson',
 'Adam Johnson',
 'Bastian Köcher',
 'Martin Donath',
 'ᗪєνιη ᗷυнʟ',
 'Christoph Junghans',
 'Raphael',
 'Major Hayden',
 'Ross A',
 'Artem Zakharchenko',
 'Nathan Rajlich',
 'Garrett D',
 'Jirka Borovec',
 'Daniel Vaz',
 'Diego Muracciole',
 'Simon Chan',
 'Carlos Alexandro',
 'Han Xiao',
 'Mike Cao',
 'Alec Thomas',
 'Dave Anderson']

In [19]:
tag_user=['p', {'class': 'f4 text-normal mb-1'}]
user= soup.find_all(tag_user)
user
f = lambda x: re.findall("            \w+",str(x) )
u = list(map(f, user ))
sp1 = []
for i in u:
    if i == [] :
        print(i)
    else:
        i = i[-1]
        sp1.append(i)

sp1        
user = []
for i in sp1:
    i = i.strip()
    user.append(i)
    
user    



[]
[]


['normanmaurer',
 'kripken',
 'mlocati',
 'hathach',
 'oblador',
 'adamchainz',
 'bkchr',
 'squidfunk',
 'onedr0p',
 'junghans',
 'glepnir',
 'major',
 'rossabaker',
 'kettanaito',
 'TooTallNate',
 'gdamore',
 'Borda',
 'dpgaspar',
 'diegomura',
 'yume',
 'caarlos0',
 'hanxiao',
 'mikecao',
 'alecthomas',
 'danderson']

#### Display the trending Python repositories in GitHub

The steps to solve this problem is similar to the previous one except that you need to find out the repository names instead of developer names.

In [25]:
print(len(names))
print(len(user))

25
25


In [28]:
for i in range(len(names)):
    print (f'{names[i]},({user[i]})' )

Norman Maurer,(normanmaurer)
Alon Zakai,(kripken)
Michele Locati,(mlocati)
Ha Thach,(hathach)
Joel Arvidsson,(oblador)
Adam Johnson,(adamchainz)
Bastian Köcher,(bkchr)
Martin Donath,(squidfunk)
ᗪєνιη ᗷυнʟ,(onedr0p)
Christoph Junghans,(junghans)
Raphael,(glepnir)
Major Hayden,(major)
Ross A,(rossabaker)
Artem Zakharchenko,(kettanaito)
Nathan Rajlich,(TooTallNate)
Garrett D,(gdamore)
Jirka Borovec,(Borda)
Daniel Vaz,(dpgaspar)
Diego Muracciole,(diegomura)
Simon Chan,(yume)
Carlos Alexandro,(caarlos0)
Han Xiao,(hanxiao)
Mike Cao,(mikecao)
Alec Thomas,(alecthomas)
Dave Anderson,(danderson)


In [11]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/python?since=daily'

In [12]:
#your code
html = requests.get(url).content
tml = requests.get(url).content
soup = BeautifulSoup(html, "lxml")
table1 = soup.find_all('h1',{'class':'h3 lh-condensed'})
print(list(table1))

[<h1 class="h3 lh-condensed">
<a data-hydro-click='{"event_type":"explore.click","payload":{"click_context":"TRENDING_REPOSITORIES_PAGE","click_target":"REPOSITORY","click_visual_representation":"REPOSITORY_NAME_HEADING","actor_id":null,"record_id":262296122,"originating_url":"https://github.com/trending/python?since=daily","user_id":null}}' data-hydro-click-hmac="935deb67be9487b4c73130e1805ddc39a9c2614659450afd6f121f845ccfc14a" href="/PaddlePaddle/PaddleOCR">
<svg aria-hidden="true" class="octicon octicon-repo mr-1 text-gray" height="16" version="1.1" viewbox="0 0 16 16" width="16"><path d="M2 2.5A2.5 2.5 0 014.5 0h8.75a.75.75 0 01.75.75v12.5a.75.75 0 01-.75.75h-2.5a.75.75 0 110-1.5h1.75v-2h-8a1 1 0 00-.714 1.7.75.75 0 01-1.072 1.05A2.495 2.495 0 012 11.5v-9zm10.5-1V9h-8c-.356 0-.694.074-1 .208V2.5a1 1 0 011-1h8zM5 12.25v3.25a.25.25 0 00.4.2l1.45-1.087a.25.25 0 01.3 0L8.6 15.7a.25.25 0 00.4-.2v-3.25a.25.25 0 00-.25-.25h-3.5a.25.25 0 00-.25.25z" fill-rule="evenodd"></path></svg>
<span 

In [13]:
f = lambda x: re.findall("        \w+\w-\w+|        \w+\w",str(x) )
u = list(map(f, table1 ))
sp1 = []
for i in u:
    if i == [] :
        print(i)
    else:
        i = i[-1]
        sp1.append(i)

sp1        
names = []
for i in sp1:
    i = i.strip()
    names.append(i)
    
names    

['PaddlePaddle',
 'AdamGold',
 'MTK-bypass',
 'apache',
 'scikit-learn',
 'facebookresearch',
 'darrenmartyn',
 'pokepetter',
 'huggingface',
 'docker',
 'public-apis',
 'zhimingshenjun',
 'tiangolo',
 'pandas-dev',
 'ansible',
 'pittcsc',
 'python',
 'Chia-Network',
 'AntixK',
 'demisto',
 'JaidedAI',
 'lyhue1991',
 'andymccurdy',
 'pytorch',
 'chipik']

#### Display all the image links from Walt Disney wikipedia page

In [30]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/Walt_Disney'

In [40]:
#your code
html = requests.get(url).content
tml = requests.get(url).content
soup = BeautifulSoup(html, "lxml")
#table1 = soup.find_all('div',{'class':'thumb tright'})
table2 = soup.find_all('a',{'class':'image'})
(list(table2))
len(table2)

23

In [48]:
f = lambda x: re.findall('href=.\W.{340}'   ,str(x) )
u = list(map(f, table2 ))
sp1 = []
for i in u:
    if i == [] :
        print(i)
    else:
        i = i[-1]
        sp1.append(i)

sp1    

img=soup.find_all('img')
src= [i['src'] for i in img]
src

['//upload.wikimedia.org/wikipedia/en/thumb/e/e7/Cscr-featured.svg/20px-Cscr-featured.svg.png',
 '//upload.wikimedia.org/wikipedia/en/thumb/8/8c/Extended-protection-shackle.svg/20px-Extended-protection-shackle.svg.png',
 '//upload.wikimedia.org/wikipedia/commons/thumb/d/df/Walt_Disney_1946.JPG/220px-Walt_Disney_1946.JPG',
 '//upload.wikimedia.org/wikipedia/commons/thumb/8/87/Walt_Disney_1942_signature.svg/150px-Walt_Disney_1942_signature.svg.png',
 '//upload.wikimedia.org/wikipedia/commons/thumb/c/c4/Walt_Disney_envelope_ca._1921.jpg/220px-Walt_Disney_envelope_ca._1921.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Newman_Laugh-O-Gram_%281921%29.webm/220px-seek%3D2-Newman_Laugh-O-Gram_%281921%29.webm.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Trolley_Troubles_poster.jpg/170px-Trolley_Troubles_poster.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/7/71/Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_of_Review_Magazine.jpg/170

In [58]:


for i in src:
    display(Image(url= f'https:{i}'))
    

In [59]:
# This is the url you will scrape in this exercise
url ='https://en.wikipedia.org/wiki/Python' 

In [97]:
html = requests.get(url).content
soup = BeautifulSoup(html, "lxml")
#table1 = soup.find_all('div',{'class':'thumb tright'})
urls=soup.find_all('a')
urls

[<a href="/browse.xhtml;jsessionid=358535E77A61F559D6CD0B5716F1FC8F"><img alt="OLRC home" src="/javax.faces.resource/u4.png.xhtml?ln=images" style="border: none;"/></a>,
 <a href="/browse.xhtml;jsessionid=358535E77A61F559D6CD0B5716F1FC8F" style="text-decoration: none; font-family: 'Calibri'; color: rgb(69, 38, 35); font-size: 15px; font-weight: bold;">
                     Home</a>,
 <a href="/browse.xhtml;jsessionid=358535E77A61F559D6CD0B5716F1FC8F"><img src="/javax.faces.resource/transparent.gif.xhtml?ln=images" style="position: absolute; top: 0px; left: 0px; width: 100%; height: 31px; cursor: pointer; border: none;"/></a>,
 <a href="/currency/currency.shtml;jsessionid=358535E77A61F559D6CD0B5716F1FC8F"><img src="/javax.faces.resource/transparent.gif.xhtml?ln=images" style="position: absolute; top: 0px; left: 0px; width: 100%; height: 31px; cursor: pointer; border: none;"/></a>,
 <a href="/classification/tables.shtml;jsessionid=358535E77A61F559D6CD0B5716F1FC8F"><img src="/javax.faces.

In [76]:
#your code
html = requests.get(url).content
soup = BeautifulSoup(html, "lxml")
#table1 = soup.find_all('div',{'class':'thumb tright'})
urls=soup.find_all('a')
urls
href=[]
for i in range(len(urls)):
    try:
        href.append(urls[i]['href'])
    except:
        pass
#urls_clean= [i['urls].replace('<a href=','') for i in urls]
href
for i in href:
    if 'https' not in i: 
        print((f'https:{i}'))
    else:
        print(url)

https:#mw-head
https:#searchInput
https://en.wikipedia.org/wiki/Python
https://en.wikipedia.org/wiki/Python
https:/wiki/Pythons
https:/wiki/Python_(genus)
https:#Computing
https:#People
https:#Roller_coasters
https:#Vehicles
https:#Weaponry
https:#Other_uses
https:#See_also
https:/w/index.php?title=Python&action=edit&section=1
https:/wiki/Python_(programming_language)
https:/wiki/CMU_Common_Lisp
https:/wiki/PERQ#PERQ_3
https:/w/index.php?title=Python&action=edit&section=2
https:/wiki/Python_of_Aenus
https:/wiki/Python_(painter)
https:/wiki/Python_of_Byzantium
https:/wiki/Python_of_Catana
https:/wiki/Python_Anghelo
https:/w/index.php?title=Python&action=edit&section=3
https:/wiki/Python_(Efteling)
https:/wiki/Python_(Busch_Gardens_Tampa_Bay)
https:/wiki/Python_(Coney_Island,_Cincinnati,_Ohio)
https:/w/index.php?title=Python&action=edit&section=4
https:/wiki/Python_(automobile_maker)
https:/wiki/Python_(Ford_prototype)
https:/w/index.php?title=Python&action=edit&section=5
https:/wiki/Pyt

#### Number of Titles that have changed in the United States Code since its last release point 

In [84]:
# This is the url you will scrape in this exercise
url = 'http://uscode.house.gov/download/download.shtml'

In [95]:
#your code
t = requests.get(url).content
soup = BeautifulSoup(t, "lxml")
#table1 = soup.find_all('div',{'class':'thumb tright'})
table = soup.find_all('div',{'class':'usctitlechanged'} )
table
#ta=soup.find_all('usctitlechanged')
#ta

[<div class="usctitlechanged" id="us/usc/t6">
 
           Title 6 - Domestic Security
 
         </div>,
 <div class="usctitlechanged" id="us/usc/t10">
 
           Title 10 - Armed Forces <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>,
 <div class="usctitlechanged" id="us/usc/t15">
 
           Title 15 - Commerce and Trade
 
         </div>,
 <div class="usctitlechanged" id="us/usc/t20">
 
           Title 20 - Education
 
         </div>,
 <div class="usctitlechanged" id="us/usc/t28">
 
           Title 28 - Judiciary and Judicial Procedure <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>,
 <div class="usctitlechanged" id="us/usc/t31">
 
           Title 31 - Money and Finance <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>,
 <div class="usctitlechanged" id="us/usc/t33">
 
           Title 33 - Navigation and Navigable Waters
 
         </div>,
 <div class="usctitlechanged" id="us/usc/t36">
 
           Title 36 - Patri

In [104]:
#f = lambda x: re.findall( '           \w+ \w+',str(x) )
#u = list(map(f, table ))
#u        
lst = [str(i.text.replace('\n','').replace('          ','').replace('        ','')) for i in table]
#lst_heroes.pop(0)
lst

['Title 6 - Domestic Security',
 'Title 10 - Armed Forces ٭',
 'Title 15 - Commerce and Trade',
 'Title 20 - Education',
 'Title 28 - Judiciary and Judicial Procedure ٭',
 'Title 31 - Money and Finance ٭',
 'Title 33 - Navigation and Navigable Waters',
 'Title 36 - Patriotic and National Observances, Ceremonies, and Organizations ٭',
 'Title 37 - Pay and Allowances of the Uniformed Services ٭',
 "Title 38 - Veterans' Benefits ٭",
 'Title 40 - Public Buildings, Property, and Works ٭',
 'Title 46 - Shipping ٭']

#### A Python list with the top ten FBI's Most Wanted names 

In [10]:
# This is the url you will scrape in this exercise
url = 'https://www.fbi.gov/wanted/topten'
fb = requests.get(url).content
soup = BeautifulSoup(fb, "lxml")
table = soup.find_all('ul',{'class':'full-grid wanted-grid-natural infinity castle-grid-block-xs-2 castle-grid-block-sm-2castle-grid-block-md-3 castle-grid-block-lg-5 dt-grid'})
#urls=soup.find_all('href')
table
lst = [i.text for i in table]

ls1 = []
for i in lst:
    if i[1] == '':
        print(i)  
    else:
        ls1.append(i[1])
        

lst
lst_clean = [i.replace("\n"," ").replace("      "," ") for i in lst]
lst_clean

[' ARNOLDO JIMENEZ   JASON DEREK BROWN   ALEXIS FLORES   JOSE RODOLFO VILLARREAL-HERNANDEZ   EUGENE PALMER   RAFAEL CARO-QUINTERO   ROBERT WILLIAM FISHER   BHADRESHKUMAR CHETANBHAI PATEL   ALEJANDRO ROSALES CASTILLO   YASER ABDEL SAID   ']

In [163]:
#your code 

####  20 latest earthquakes info (date, time, latitude, longitude and region name) by the EMSC as a pandas dataframe

In [20]:
# This is the url you will scrape in this exercise
url = 'https://www.emsc-csem.org/Earthquake/'

In [36]:
#your code

#your code
t = requests.get(url).content
soup = BeautifulSoup(t, "lxml")
#table1 = soup.find_all('div',{'class':'thumb tright'})
table = soup.find_all('tbody',{'id':'tbody'} )[0]
table
#ta=soup.find_all('usctitlechanged')
#ta
#b = table.find_all('tbody')
#b
table = soup.find('tbody')
rows = [row for row in table.find_all('tr',{'onclick':True})][:20]
data1 = [data.find_all('td')[3:-1] for data in rows]
data_date = [data1[i][0].find('a').text for i in range(len(data1))]
data_ndate_txt = [[element.text for element in row[1:]] for row in data1]
# data_ndate_txt
data_txt = data1.copy()
for i in range(len(data_ndate_txt)):
#     data_txt[i][0] = data_date[i]
#     data_txt[][] = data_ndate_txt[]
    data_ndate_txt[i].insert(0,data_date[i])
df = pd.DataFrame(data_ndate_txt,index = list(range(1,21)))

df.drop(columns = 2, axis=1, inplace = True)
df.drop(columns = 4, axis=1, inplace = True)
df.drop(columns = 6, axis=1, inplace = True)
 
df_new = df.rename(columns={0: 'Time',1: 'Latitud',3: 'Longitud',5: 'Depth',7: 'Mag',8: 'Region'})
df_new.head()


Time Latitud Longitud Depth  Mag  \
1  2021-01-28   17:06:31.8  41.28    49.83     80  4.6   
2  2021-01-28   16:53:22.0  15.57    97.74     16  4.1   
3  2021-01-28   16:32:41.0   5.03   127.17    127  3.1   
4  2021-01-28   16:17:25.9  39.41   175.15     79  3.6   
5  2021-01-28   16:16:41.1  37.25    33.23      7  2.1   

                              Region  
1   CASPIAN SEA, OFFSHORE AZERBAIJAN  
2            OFFSHORE OAXACA, MEXICO  
3          PHILIPPINE ISLANDS REGION  
4        NORTH ISLAND OF NEW ZEALAND  
5                     CENTRAL TURKEY

#### Display the date, days, title, city, country of next 25 hackathon events as a Pandas dataframe table

<body><h1>hackevents.co</h1><p>Coming soon.</p></body>

#### Count number of tweets by a given Twitter account.

You will need to include a ***try/except block*** for account names not found. 
<br>***Hint:*** the program should count the number of tweets for any provided account

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

#### Number of followers of a given twitter account

You will need to include a ***try/except block*** in case account/s name not found. 
<br>***Hint:*** the program should count the followers for any provided account

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
#your code

#### List all language names and number of related articles in the order they appear in wikipedia.org

In [70]:
# This is the url you will scrape in this exercise
url = 'https://www.wikipedia.org/'

t = requests.get(url).content
soup = BeautifulSoup(t, "lxml")
table = soup.find_all('div',{'class':'central-featured'} )
table
table = soup.find_all('a')
table = soup.find_all('div',{'dir':'ltr'} )
table
for name in table:
    print(name.get_text())



English
6 235 000+ articles




日本語
1 250 000+ 記事




Deutsch
2 528 000+ Artikel




Español
1 655 000+ artículos




Русский
1 693 000+ статей




Français
2 292 000+ articles




Italiano
1 669 000+ voci




中文
1 172 000+ 條目




Polski
1 452 000+ haseł




Português
1 053 000+ artigos




#### A list with the different kind of datasets available in data.gov.uk 

In [87]:
# This is the url you will scrape in this exercise
url = 'https://data.gov.uk/'
t = requests.get(url).content
soup = BeautifulSoup(t, "lxml")
table = list(soup.find_all('a',{'class':'govuk-link'} ))[3:]



for name in table:
    print(name.get_text())





Business and economy
Crime and justice
Defence
Education
Environment
Government
Government spending
Health
Mapping
Society
Towns and cities
Transport


In [ ]:
#your code 
t = requests.get(url).content
soup = BeautifulSoup(t, "lxml")
table = soup.find_all('div',{'class':'central-featured'} )
table
table = soup.find_all('a')
table = soup.find_all('div',{'dir':'ltr'} )
table
for name in table:
    print(name.get_text())

#### Top 10 languages by number of native speakers stored in a Pandas Dataframe

In [138]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/List_of_languages_by_number_of_native_speakers'
t = requests.get(url).content
soup = BeautifulSoup(t, "lxml")
table = soup.find_all('a')[46:108]

text = [i.text for i in table]

text.remove('[9]')
text.remove('[10]')

text


['Mandarin Chinese',
 'Sino-Tibetan',
 'Sinitic',
 'Spanish',
 'Indo-European',
 'Romance',
 'English',
 'Indo-European',
 'Germanic',
 'Hindi',
 'Hindustani',
 'Indo-European',
 'Indo-Aryan',
 'Bengali',
 'Indo-European',
 'Indo-Aryan',
 'Portuguese',
 'Indo-European',
 'Romance',
 'Russian',
 'Indo-European',
 'Balto-Slavic',
 'Japanese',
 'Japonic',
 'Japanese',
 'Western Punjabi',
 'Indo-European',
 'Indo-Aryan',
 'Marathi',
 'Indo-European',
 'Indo-Aryan',
 'Telugu',
 'Dravidian',
 'South-Central',
 'Wu Chinese',
 'Sino-Tibetan',
 'Sinitic',
 'Turkish',
 'Turkic',
 'Oghuz',
 'Korean',
 'Koreanic',
 'language isolate',
 'French',
 'Indo-European',
 'Romance',
 'German',
 'Standard German',
 'Indo-European',
 'Germanic',
 'Vietnamese',
 'Austroasiatic',
 'Vietic',
 'Tamil',
 'Dravidian',
 'South',
 'Yue Chinese',
 'Sino-Tibetan',
 'Sinitic',
 'Urdu']

In [135]:
lst_lst  = []
for i in range(0, len(text), 3):
     lst_lst.append(text[i])
lst_lst    

['Mandarin Chinese',
 'Spanish',
 'English',
 'Hindi',
 'Indo-Aryan',
 'Indo-Aryan',
 'Romance',
 'Balto-Slavic',
 'Japanese',
 'Indo-Aryan',
 'Indo-Aryan',
 'South-Central',
 'Sinitic',
 'Oghuz',
 'language isolate',
 'Romance',
 'Indo-European',
 'Austroasiatic',
 'Dravidian',
 'Sino-Tibetan']

In [158]:
#your code
req = requests.get(url)
sopa = BeautifulSoup(req.content, 'html')
a=sopa.find_all('a')
lg=[i.text for i in a]
lg=lg[46:77]

lg.remove('[9]')
lg.remove('[10]')
lg.remove('Sino-Tibetan')
lg.remove('Sinitic')
lg.remove('Indo-European')
lg.remove('Romance')
lg.remove('Germanic')
lg.remove('Indo-European')
lg.remove('Hindustani')
lg.remove('Indo-European')
lg.remove('Indo-Aryan')
lg.remove('Indo-European')
lg.remove('Indo-Aryan')
lg.remove('Indo-European')
lg.remove('Romance')
lg.remove('Indo-European')
lg.remove('Balto-Slavic')
lg.remove('Japonic')
lg.remove('Japanese')
lg.remove('Indo-European')
lg.remove('Indo-Aryan')

lg

['Mandarin Chinese',
 'Spanish',
 'English',
 'Hindi',
 'Bengali',
 'Portuguese',
 'Russian',
 'Japanese',
 'Western Punjabi',
 'Marathi']

In [183]:

table = sopa.find('table', {'class' :'wikitable sortable'})
headers = [head.text.replace(' \n','') for head in table.find_all('th')]
headers
body = [bod.text.replace(' \n' ,'') for bod in table.find_all('td')]
body

['1\n',
 'Mandarin Chinese\n',
 '918\n',
 '11.922%\n',
 'Sino-Tibetan\n',
 'Sinitic\n',
 '2\n',
 'Spanish\n',
 '480\n',
 '5.994%\n',
 'Indo-European\n',
 'Romance\n',
 '3\n',
 'English\n',
 '379\n',
 '4.922%\n',
 'Indo-European\n',
 'Germanic\n',
 '4\n',
 'Hindi (sanskritised Hindustani)[9]\n',
 '341\n',
 '4.429%\n',
 'Indo-European\n',
 'Indo-Aryan\n',
 '5\n',
 'Bengali\n',
 '228\n',
 '2.961%\n',
 'Indo-European\n',
 'Indo-Aryan\n',
 '6\n',
 'Portuguese\n',
 '221\n',
 '2.870%\n',
 'Indo-European\n',
 'Romance\n',
 '7\n',
 'Russian\n',
 '154\n',
 '2.000%\n',
 'Indo-European\n',
 'Balto-Slavic\n',
 '8\n',
 'Japanese\n',
 '128\n',
 '1.662%\n',
 'Japonic\n',
 'Japanese\n',
 '9\n',
 'Western Punjabi[10]\n',
 '92.7\n',
 '1.204%\n',
 'Indo-European\n',
 'Indo-Aryan\n',
 '10\n',
 'Marathi\n',
 '83.1\n',
 '1.079%\n',
 'Indo-European\n',
 'Indo-Aryan\n',
 '11\n',
 'Telugu\n',
 '82.0\n',
 '1.065%\n',
 'Dravidian\n',
 'South-Central\n',
 '12\n',
 'Wu Chinese\n',
 '81.4\n',
 '1.057%\n',
 'Sino-Tib

In [184]:
j = 0
lst_lst = []
for i in range(0, len(body),6):
    lst_lst.append(body[j:i])
    j = i
body = lst_lst[1:11]
body

[['1\n',
  'Mandarin Chinese\n',
  '918\n',
  '11.922%\n',
  'Sino-Tibetan\n',
  'Sinitic\n'],
 ['2\n', 'Spanish\n', '480\n', '5.994%\n', 'Indo-European\n', 'Romance\n'],
 ['3\n', 'English\n', '379\n', '4.922%\n', 'Indo-European\n', 'Germanic\n'],
 ['4\n',
  'Hindi (sanskritised Hindustani)[9]\n',
  '341\n',
  '4.429%\n',
  'Indo-European\n',
  'Indo-Aryan\n'],
 ['5\n', 'Bengali\n', '228\n', '2.961%\n', 'Indo-European\n', 'Indo-Aryan\n'],
 ['6\n', 'Portuguese\n', '221\n', '2.870%\n', 'Indo-European\n', 'Romance\n'],
 ['7\n',
  'Russian\n',
  '154\n',
  '2.000%\n',
  'Indo-European\n',
  'Balto-Slavic\n'],
 ['8\n', 'Japanese\n', '128\n', '1.662%\n', 'Japonic\n', 'Japanese\n'],
 ['9\n',
  'Western Punjabi[10]\n',
  '92.7\n',
  '1.204%\n',
  'Indo-European\n',
  'Indo-Aryan\n'],
 ['10\n',
  'Marathi\n',
  '83.1\n',
  '1.079%\n',
  'Indo-European\n',
  'Indo-Aryan\n']]

In [189]:
df = pd.DataFrame(body ,columns = headers)
df.columns = [col.replace("\n", "") for col in df.columns]
df
df["Language"] = df["Language"].replace({'\n':''}, regex=True)
df["Language"] = df["Language"].replace({'[10]' :''}, regex=True)
df["Language"] = df["Language"].replace({'[9]' :''}, regex=True)

df.drop(columns = 'Rank', axis=1, inplace = True)
df.drop(columns = 'Percentageof world pop.(March 2019)[8]', axis=1, inplace = True)
df.drop(columns = 'Language family', axis=1, inplace = True)
df.drop(columns = 'Branch', axis=1, inplace = True)

df["Speakers(millions)"] = df["Speakers(millions)"].replace({'\n':''}, regex=True)

df

Language Speakers(millions)
0                   Mandarin Chinese                918
1                            Spanish                480
2                            English                379
3  Hindi (sanskritised Hindustani)[]                341
4                            Bengali                228
5                         Portuguese                221
6                            Russian                154
7                           Japanese                128
8                  Western Punjabi[]               92.7
9                            Marathi               83.1

### BONUS QUESTIONS

#### Scrape a certain number of tweets of a given Twitter account.

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
# your code

#### IMDB's Top 250 data (movie name, Initial release, director name and stars) as a pandas dataframe

In [ ]:
# This is the url you will scrape in this exercise 
url = 'https://www.imdb.com/chart/top'

In [ ]:
# your code

#### Movie name, year and a brief summary of the top 10 random movies (IMDB) as a pandas dataframe.

In [ ]:
#This is the url you will scrape in this exercise
url = 'http://www.imdb.com/chart/top'

In [ ]:
#your code

#### Find the live weather report (temperature, wind speed, description and weather) of a given city.

In [ ]:
#https://openweathermap.org/current
city = city=input('Enter the city:')
url = 'http://api.openweathermap.org/data/2.5/weather?'+'q='+city+'&APPID=b35975e18dc93725acb092f7272cc6b8&units=metric'

In [ ]:
# your code

#### Book name,price and stock availability as a pandas dataframe.

In [ ]:
# This is the url you will scrape in this exercise. 
# It is a fictional bookstore created to be scraped. 
url = 'http://books.toscrape.com/'

In [ ]:
#your code